orttraining/orttraining/test/training_api/common/synthetic_data_loader.cc

```diff
@@ -73,50 +97,41 @@ bool SyntheticDataLoader::GetNextSampleBatch(std::vector<OrtValue*>& batches) {
-    auto shape_vector = input_ptr->ShapeVector();
-    // Be noted: the created OrtValue won't clean the raw data after its lifetime ended.
-    auto ptr_flt = dynamic_cast<TypedSyntheticInput<float>*>(input_ptr);
-    if (ptr_flt) {
-      OrtValue* value = NULL;
-      ORT_RETURN_ON_ERROR(ort_api->CreateTensorWithDataAsOrtValue(memory_info,
-      input_ptr->GetData<float>().data(), (input_ptr->NumOfElements() * sizeof(float)),
-      shape_vector.data(), shape_vector.size(),
-      ONNXTensorElementDataType::ONNX_TENSOR_ELEMENT_DATA_TYPE_FLOAT,
-      &value));
-      batches.emplace_back(value);
-      continue;
-    }
-
-    auto ptr_int = dynamic_cast<TypedSyntheticInput<int64_t>*>(input_ptr);
-    if (ptr_int) {
-      OrtValue* value = NULL;
-      ORT_RETURN_ON_ERROR(ort_api->CreateTensorWithDataAsOrtValue(memory_info,
-      input_ptr->GetData<int64_t>().data(), (input_ptr->NumOfElements() * sizeof(int64_t)),
-      shape_vector.data(), shape_vector.size(),
-      ONNXTensorElementDataType::ONNX_TENSOR_ELEMENT_DATA_TYPE_INT64,
-      &value));
-      batches.emplace_back(value);
-      continue;
-    }
-
-    auto ptr_int32 = dynamic_cast<TypedSyntheticInput<int32_t>*>(input_ptr);
-    if (ptr_int32) {
+  for (size_t i = 0; i < sample.NumOfInput(); ++i) {
+    SyntheticInput& input = sample.GetInputAtIndex(i);
+
+    std::visit([&batches, &input, &ort_api, &memory_info](auto&& arg) -> void {
+      ONNXTensorElementDataType elem_data_type;
+      using T = std::decay_t<decltype(arg)>;
+      if constexpr (std::is_same_v<T, std::vector<float>>) {
+        elem_data_type = Ort::TypeToTensorType<float>::type;
+      } else if constexpr (std::is_same_v<T, std::vector<int32_t>>) {
+        elem_data_type = Ort::TypeToTensorType<int32_t>::type;
+      } else if constexpr (std::is_same_v<T, std::vector<int64_t>>) {
+        elem_data_type = Ort::TypeToTensorType<int64_t>::type;
+      } else if constexpr (std::is_same_v<T, std::vector<uint8_t>>) {
+        elem_data_type = ONNXTensorElementDataType::ONNX_TENSOR_ELEMENT_DATA_TYPE_BOOL;
+      } else {
+        throw std::runtime_error("Unsupported element type.");
+      }
+
+      void* p_data = arg.data();
       OrtValue* value = nullptr;
-      ORT_RETURN_ON_ERROR(ort_api->CreateTensorWithDataAsOrtValue(memory_info,
-      input_ptr->GetData<int32_t>().data(), (input_ptr->NumOfElements() * sizeof(int32_t)),
-      shape_vector.data(), shape_vector.size(),
-      ONNXTensorElementDataType::ONNX_TENSOR_ELEMENT_DATA_TYPE_INT32,
-      &value));
-      batches.emplace_back(value);
-      continue;
-    }
+      auto shape_vector = input.ShapeVector();
+      // Be noted: the created OrtValue won't clean the raw data after its lifetime ended.
+      ORT_RETURN_ON_ERROR(ort_api->CreateTensorWithDataAsOrtValue(
+          memory_info,
+          p_data, (input.NumOfElements() * sizeof(T)),
+          shape_vector.data(), shape_vector.size(),
+          elem_data_type,
+          &value));
```

Comment:
Use C++ API.

Reply:
I would prefer to align with the usage in other places, we can change to C++ later if customer need use and we have implemented those needed C++ training API.